### ISSUE TITLE
student_panel: a barely-loaded period counts as fully observed, advancing the label window one cycle too early

### ISSUE THREAD

**Issue by MatiasLoiseau** [MatiasLoiseau/PREDUN#7]
## Summary

`marts.student_panel` closes its academic calendar at the period containing `max(fecha)`:

```sql
-- student_panel.sql:81-83
data_bounds as (
    select max(period_start) as last_period_start
    from cursadas_dated
),

-- student_panel.sql:93-102
periods as (
    select ...
    from generate_series('2011-03-01'::date, '2035-09-01'::date, interval '6 months') as gs(period_start)
    cross join data_bounds db
    where gs.period_start <= db.last_period_start
),
```

This asks **whether any acta exists** in a period, never **how many**. One record is enough for the period to enter the calendar as if it were a normal term.

That matters because every institutional delivery arrives *before* its own term's actas are closed. Grades are recorded four to five months after the student sits the course:

| Month | Share of all actas |
|---|---:|
| July | **44.2%** |
| December | **36.4%** |
| November | 9.3% |
| June | 6.9% |
| Everything else | 3.2% |

So `1C` closes in June–July and `2C` closes in November–December, while deliveries land in May, August and September.

## What each delivery actually contained

Acta counts per academic period, per delivery, from `staging.cursada_historica_raw` (raw rows, before dedup and before the 29-career filter):

| Delivery | `max(FECHA)` | period `2024_2C` | period `2025_1C` |
|---|---|---:|---:|
| `2024_2C` | 2024-09-10 | **456** | — |
| `2025_1C` | 2025-05-12 | 62,180 | **1,604** |
| `2025_2C` | 2025-08-04 | 62,113 | **65,349** |

- `2024_2C` arrives 10 days into the second term. Its actas close in November. The period is **0.7% loaded**.
- `2025_1C` arrives mid-first-term. Its actas close in July. The period is **2.5% loaded**.
- `2025_2C` arrives on 4 August, *after* the `1C` actas closed and *before* the `2C` ones start. It is the only one of the three whose last period is complete.

In the first two cases the panel gains a phantom period, and because `VAL_PERIOD` is the most recent labelable period and `TRAIN_CUTOFF = VAL_PERIOD - 4`, the whole evaluation window slides forward one cycle too early.

## Consequence for the labels

`dropout_next` is 1 when the next four periods show no activity. If the fourth of those periods is empty because its actas have not been closed yet, students who did come back are recorded as dropouts.

Measured on the `2023_1C` validation set (23,685 at-risk labelled rows), counting students whose only activity inside the label window fell in `2025_1C`:

| | |
|---|---:|
| Rows in the validation set | 23,685 |
| Labelled 1 with complete data (delivery `2025_2C`) | 10,319 |
| **Would have been labelled 1 with `2025_1C` empty (delivery `2025_1C`)** | **+222** |

Validation prevalence moves from **0.4357 to 0.4451**.

### Independent confirmation

`predictions.drift_metrics` holds `dropout_next` for both cycles with an **identical row count** and a different mean:

| Cycle | reference | n_current | mean_current |
|---|---|---:|---:|
| `2025_1C` | `2021_1C` | 91,068 | **0.415** |
| `2025_2C` | `2021_1C` | 91,068 | **0.413** |

The gap is 0.002, and 222 / 91,068 = 0.0024. The two measurements agree.

The reference means are identical in both cycles (0.370), which is expected: labels at or before `2021_1C` close their window at `2023_1C` and cannot be touched by 2025 data. Only the post-cutoff side moves.

### Effect on the reported per-cycle metric

From `predictions.model_evaluations` and the MLflow backend (`winning_model=true`):

| Delivery | `train_period_cutoff` | Validation period | train_size | val_size | AUC |
|---|---|---|---:|---:|---:|
| `2024_2C` | `2020_2C` | `2022_2C` | 192,931 | 23,044 | 0.9121 |
| `2025_1C` | `2021_1C` | `2023_1C` | 213,107 | 23,685 | **0.9329** |
| `2025_2C` | `2021_1C` | `2023_1C` | 213,107 | 23,685 | **0.9304** |

Cycles `2025_1C` and `2025_2C` train and validate on exactly the same periods and the same row counts, yet the AUC differs by 0.0025. That difference is not model variance — it is 222 corrupted labels in the earlier cycle.

Had the calendar required a loaded period, the three cycles would have walked forward one term at a time:

| Delivery | Validation as-is | Validation as it should be |
|---|---|---|
| `2024_2C` | `2022_2C` | `2022_1C` |
| `2025_1C` | `2023_1C` | `2022_2C` |
| `2025_2C` | `2023_1C` | `2023_1C` |

## Proposed fix

Require a period to carry an acta volume comparable to its predecessors before it closes the calendar.

```diff
+-- Acta volume per period, with the mean of the four preceding periods.
+period_volume as (
+    select
+        period_start,
+        count(*) as n_actas,
+        avg(count(*)) over (
+            order by period_start
+            rows between 4 preceding and 1 preceding
+        ) as avg_prev
+    from cursadas_dated
+    group by 1
+),
+
 data_bounds as (
-    select max(period_start) as last_period_start
-    from cursadas_dated
+    -- A period only closes the calendar once its actas are actually loaded.
+    -- Deliveries arrive before their own term closes (actas land in Jul and Dec),
+    -- so the newest period can hold a handful of records and still be counted
+    -- as fully observed, sliding VAL_PERIOD forward one cycle too early.
+    select max(period_start) as last_period_start
+    from period_volume
+    where avg_prev is null              -- first periods of the series
+       or n_actas >= 0.5 * avg_prev
 ),
```

`percentile_cont` is an ordered-set aggregate and cannot be used as a window function in Postgres, hence the mean rather than the median.

### Why 0.5

Acta volume per period on the current `canonical.cursada_historica` (deduplicated, 29 careers), against the mean of the four preceding periods:

| Period | Actas | Mean of prev. 4 | % of mean |
|---|---:|---:|---:|
| `2025_1C` | 45,084 | 41,358 | 109% |
| `2024_2C` | 41,534 | 40,487 | 103% |
| `2024_1C` | 41,307 | 41,355 | 100% |
| `2023_2C` | 40,334 | 41,812 | 96% |
| `2023_1C` | 42,255 | 42,371 | 100% |
| `2022_2C` | 38,052 | 43,430 | 88% |
| `2022_1C` | 44,780 | 41,347 | 108% |
| `2021_2C` | 42,160 | 40,313 | 105% |

Legitimate periods sit between **88% and 115%** across the whole series. The two artifact periods held **0.7%** and **2.5%** of the volume they ended up with once their actas closed. A 0.5 cut leaves a wide margin on both sides and no legitimate period comes near it. Same spirit as the 20% threshold already calibrated in `predun_dbt/tests/assert_no_count_inflation.sql`.

## Post-fix verification

1. Rebuilt from the `2025_2C` delivery, `student_panel` should be unchanged: 687,186 rows, 394,362 at risk, 304,175 labelled, last period `2025_1C`.
2. `VAL_PERIOD` should stay `2023_1C` and `TRAIN_CUTOFF` `2021_1C` for this delivery, so `drift_train_predict` reproduces AUC 0.9304.
3. Rebuilding from a delivery truncated at 2025-05-12 should now yield `VAL_PERIOD = 2022_2C`, not `2023_1C`.
4. A dbt test asserting that the last calendar period holds at least half the mean volume of the preceding four would make the invariant explicit.

## Scope

**Affected:** the label window selection for any delivery that lands before its own term's actas close, which is `2024_2C` and `2025_1C`. Their reported per-cycle AUC is optimistic by roughly 0.0025.

**Not affected:** the panel currently in the database and everything computed from it. The `2025_2C` delivery arrived with `2025_1C` fully closed, so the backtest, the ablation, the paired test, the fairness analysis and the drift table all run over complete label windows.

**Related, not covered here:** `marts.student_status` derives its cutoff from the same `max(fecha)` (`max(fecha) - interval '2 years'`). A delivery arriving mid-term shifts that cutoff by a few months and therefore the `estudiando` cohort, which is a smaller and separate effect.

This is the label-side counterpart of the partially-loaded-period effect already noted for scoring in #5. The two act in opposite directions: the scoring path depressed probabilities, this one inflates the positive class.

**Environment:** `dagster-predun` / Postgres 15. **Verified at commit** `a650ab9`.
